src/main/java/services/MenuSessionFactory.java

```diff
@@ -0,0 +1,102 @@
+package services;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.session.SessionFrame;
+import org.commcare.suite.model.EntityDatum;
+import org.commcare.suite.model.MenuDisplayable;
+import org.commcare.suite.model.SessionDatum;
+import org.commcare.suite.model.StackFrameStep;
+import org.commcare.util.screen.CommCareSessionException;
+import org.commcare.util.screen.EntityScreen;
+import org.commcare.util.screen.MenuScreen;
+import org.commcare.util.screen.Screen;
+import org.javarosa.core.model.actions.FormSendCalloutHandler;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import repo.FormSessionRepo;
+import repo.SerializableMenuSession;
+import session.MenuSession;
+
+import java.util.Hashtable;
+import java.util.Set;
+import java.util.Stack;
+import java.util.Vector;
+
+/**
+ * Class containing logic for accepting a NewSessionRequest and services,
+ * restoring the user, opening the new form, and returning the question list response.
+ */
+@Component
+public class MenuSessionFactory {
+
+    @Autowired
+    private RestoreFactory restoreFactory;
+
+    @Autowired
+    private InstallService installService;
+
+    @Value("${commcarehq.host}")
+    private String host;
+
+    private static final Log log = LogFactory.getLog(MenuSessionFactory.class);
+
+    /**
+     * Rebuild the MenuSession from its stack frame. This is used after end of form navigation.
+     * By re-walking the frame, we establish the set of selections the user 'would' have made to get
+     * to this state without doing end of form navigation. Such a path must always exist in a valid app.
+    */
+    public void rebuildSessionFromFrame(MenuSession menuSession) {
+        Vector<StackFrameStep> steps = menuSession.getSessionWrapper().getFrame().getSteps();
+        menuSession.resetSession();
+        Screen screen = menuSession.getNextScreen();
+        while (screen != null) {
+            String currentStep = null;
+            if (screen instanceof MenuScreen) {
+                MenuDisplayable[] options = ((MenuScreen) screen).getMenuDisplayables();
+                for (int i = 0; i < options.length; i++) {
+                    for (StackFrameStep step : steps) {
+                        if (step.getId().equals(options[i].getCommandID())) {
+                            currentStep = String.valueOf(i);
+                        }
+                    }
+                }
+            } else if (screen instanceof EntityScreen) {
+                EntityScreen entityScreen = (EntityScreen) screen;
+                SessionDatum neededDatum = entityScreen.getSession().getNeededDatum();
+                Set<String> caseIds = entityScreen.getReferenceMap().keySet();
```

Comment:
noting that "caseIds" is a misnomer here.

Reply:
@ctsims what should it be?

Reply:
Also wondering this - would you prefer `entityIds` ?

Reply:
entityIds is better, yeah.